sklearn/ensemble/_hist_gradient_boosting/_predictor_bitset.pyx

```diff
@@ -0,0 +1,104 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: language_level=3
+# cython: nonecheck=False
+# distutils: language=c++
+
+from ._bitset cimport in_bitset
+from .common cimport BITSET_INNER_DTYPE_C
+from libc.limits cimport CHAR_BIT
+from libcpp.vector cimport vector
+
+
+cdef inline unsigned char in_vec_bitset(vector[BITSET_INNER_DTYPE_C]& bitset,
+                                        int value) nogil:
+    cdef:
+        unsigned int i1 = value // 32
+        unsigned int i2 = value % 32
+
+    if bitset.size() < i1 + 1:
+        return 0
+    return (bitset[i1] >> i2) & 1
+
+
+cdef inline void insert_vec_bitset(vector[BITSET_INNER_DTYPE_C]& bitset,  # OUT
+                                   int value) nogil:
+    cdef:
+        unsigned int i1 = value // 32
+        unsigned int i2 = value % 32
+
+    if bitset.size() < i1 + 1:
+        bitset.resize(i1 + 1, 0)
+    bitset[i1] |= (1 << i2)
+
+
+cdef class PredictorBitSet:
+    def __init__(self, list bin_thresholds,
+                 const unsigned char [:] is_categorical):
+        """Creates bitset for all known categories"""
+        if is_categorical is None or bin_thresholds is None:
+            return
+
+        cdef:
+            int i
+            X_DTYPE_C category
+            unsigned int i1
+            unsigned int i2
+
+        for i in range(is_categorical.shape[0]):
+            if is_categorical[i] == 0:
+                continue
+            for category in bin_thresholds[i]:
+                insert_vec_bitset(self.feature_idx_raw_cats[i],
+                                  <int>(category))
+
+    def insert_categories_bitset(self, unsigned int node_idx,
+                                 X_DTYPE_C[:] category_bins,
+                                 BITSET_INNER_DTYPE_C[:] cat_bitset):
+        """Insert category into bitset for raw categories and binned cateogires
+        for node_idx.
+        """
+        cdef:
+            BITSET_INNER_DTYPE_C val
+            int k, offset
+            int cardinality = category_bins.shape[0]
+            int BITSET_SIZE = sizeof(BITSET_INNER_DTYPE_C) * CHAR_BIT
+            unsigned int i1, i2
+
+        self.node_to_binned_bitset[node_idx].resize(cat_bitset.shape[0])
+        for k in range(cat_bitset.shape[0]):
+            offset = BITSET_SIZE * k
+            val = cat_bitset[k]
+            self.node_to_binned_bitset[node_idx][k] = val
+            while val and offset < cardinality:
+                if val % 2:
```

Comment:
I think it'd be clearer to stick to the previous use of the masks: `val & 1`. Took me a while to realize what this was doing.